docs/utxo-hd/v0.1/for-developers.md

```diff
@@ -0,0 +1,456 @@
+
+### The mempool
+
+The mempool behaves pretty much as a virtual block. The design is not
+particularly complex as we just ask the backing store for the inputs for a
+transaction when applying it. The only caveat compared to the pre-UTxO-HD
+implementation is that previously it cached the latest ledger state and
+therefore we could run a separate thread that would sync the mempool with the
+LedgerDB and revalidate all the transactions when necessary, but now we might
+not be able to apply a transaction if the ledger state on top of which we had
+applied the others is now gone from the LedgerDB. Therefore, adding a
+transaction might in some cases trigger a sync with the LedgerDB and therefore a
+revalidation of the previous transactions.
+
+### Ledger state queries
+
+Most of the queries don't require the UTxO set but there are three in particular
+that do: `GetUTxOByTxIn`, `GetUTxOWhole` and `GetUTxOByAddress`. We assume that
+`GetUTxOWhole` is a debug query so we don't pay much attention to it as we
+expect it will be used with small-ish UTxO sets. For `GetUTxOByTxIn`, the story
+is simple, query the Backing store for the `TxIn` and forward it through the
+sequence of diffs.
+
+However, it is `GetUTxOByAddress` that poses a real problem, as we need to query
+the whole UTxO set, apply all the differences to it and then traverse it
+entirely to find out the UTxOs belonging to an address. This query is now quite
+slow and in fact its usage is discouraged. It should not be a responsibility of
+the node to maintain access to this if it is not needed by the logic that runs
+the blockchain, so the plan is to move this into a separate process/client that
+runs an index of UTxOs by address that can provide fast access to it (see [#4678](https://github.com/input-output-hk/cardano-node/issues/4678)).
+
+Currently, a pre-UTxO-HD node performs this query in 7 seconds, but our ad-hoc
+measurements have shown that a UTxO-HD node takes ~40 seconds when using the
+InMemory backend and ~90 seconds when using the LMDB backend.
+
+### The hard-fork tables
+
+It is worth mentioning one important caveat that has given us quite a lot of
+headaches while developing UTxO-HD. When a blockchain is running a
+`HardForkBlock` (as Cardano does), the overall type of the data stored in the
+`DbChangelog` and in the `BackingStore` are maps from some key type into an
+N-ary sum of value types (one for each era). 
+
+But applying a block in a particular era requires a ledger state in said era.
+Whenever we inject `HardForkBlock` ledger tables into the ledger tables of the
+particular era, we translate the inputs that belong to previous eras and we
+unwrap the N-ary sum value effectively duplicating the map in which they are
+contained.
+
+This is fine in the Consensus use case because the set of inputs for each block
+will be quite small. However this imposes an important restriction, one cannot
+just simply project in and out the whole UTxO set on every ledger state (and not
```

Comment:
Yeah, I think it should be clear from the whole sentence. If one does a naive integration and just says "ok, you want me to use these tables, so I will just project in and out the whole UTxO set so that I conform to the types", then the memory will blow up. I think it is conveyed by the sentence 

> one cannot just naively project in and out a whole UTxO set on every ledger state (and not mantain a sequence of differences) to do a preliminary integration of UTxO-HD